File: 07.py
with open("7.txt") as f:
    crabs = [int(x) for x in f.readline().strip().split(',')]
#crabs = [16,1,2,0,4,2,7,1,2,14]

mincrab = min(crabs)
maxcrab = max(crabs)

costs = dict()
for i in range(mincrab, maxcrab+1):
    costs[i] = sum([abs(x - i) for x in crabs])

minpos = min(costs, key=costs.get)
print('part1', costs[minpos])

File: 07.clj
(load-file "utils.clj")

(def crabs
  (map
    utils/parseInt
    (clojure.string/split (first (utils/read-lines "7.txt")) #",")))
    
;(def crabs [16 1 2 0 4 2 7 1 2 14])

(defn- abs [n] (max n (-' n)))

(defn do-part1 [crabs]
  (apply min
    (let [mincrab (apply min crabs)
          maxcrab (apply max crabs)]
      (into []
        (for [c (range mincrab (inc maxcrab))]
          (->> crabs
            (map #(abs (- %1 c)))
            (reduce +)))))))

(defn do-part2 [crabs]
  (letfn [
    (calc [i x]
      (let [diff (abs (- x i))]
        (as-> (inc diff) $ 
          (* diff $)
          (quot $ 2))))]
    (apply min
      (let [mincrab (apply min crabs)
            maxcrab (apply max crabs)]
        (into []
          (for [c (range mincrab (inc maxcrab))]
            (->> crabs
              (map (partial calc c))
              (reduce +))))))))

(println "part1" (do-part1 crabs))
(println "part2" (do-part2 crabs))